src/parquet/bloom_filter.cc

```diff
@@ -0,0 +1,164 @@
+// specific language governing permissions and limitations
+// under the License.
+
+
+
+#include <algorithm>
+#include <cmath>
+#include <cstdint>
+
+#include "arrow/status.h"
+#include "arrow/util/bit-util.h"
+#include "parquet/bloom_filter.h"
+#include "parquet/bloom_filter_algorithm.h"
+#include "parquet/exception.h"
+#include "parquet/murmur3.h"
+#include "parquet/util/logging.h"
+
+namespace parquet {
+
+BloomFilter::BloomFilter(uint32_t num_bytes)
+    : pool_(::arrow::default_memory_pool()),
+	  bitset_(NULL),
+	  is_internal_bitset_(true),
+      num_bytes_(num_bytes),
+      hash_strategy_(HashStrategy::MURMUR3_X64_128),
+      algorithm_(Algorithm::BLOCK) {
+  switch (hash_strategy_) {
+    case HashStrategy::MURMUR3_X64_128:
+      this->hasher_.reset(new MurmurHash3());
+      break;
+    default:
+      throw parquet::ParquetException("Unsupported hash strategy.");
+  }
+
+  switch (algorithm_) {
+  case Algorithm::BLOCK:
+	  this->bloom_algorithm_.reset(new BlockBasedAlgorithm());
+	  break;
+  default:
+	  throw parquet::ParquetException("Unsupported Bloom filter algorithm");
+  }
+
+  InitBitset(num_bytes);
+}
+
+void BloomFilter::InitBitset(uint32_t num_bytes) {
+  if (num_bytes < MINIMUM_BLOOM_FILTER_BYTES) {
+    num_bytes = MINIMUM_BLOOM_FILTER_BYTES;
+  }
+
+  // Get next power of 2 if it is not power of 2.
+  if ((num_bytes & (num_bytes - 1)) != 0) {
+    num_bytes = static_cast<uint32_t>(::arrow::BitUtil::NextPower2(num_bytes));
+  }
+
+  if (num_bytes > MAXIMUM_BLOOM_FILTER_BYTES) {
+	  num_bytes = MAXIMUM_BLOOM_FILTER_BYTES;
+  }
+
+  ::arrow::Status status = pool_->Allocate(num_bytes,
+                                           reinterpret_cast<uint8_t**>(&bitset_));
+  if (!status.ok()) {
+    throw parquet::ParquetException("Failed to allocate buffer for bitset");
+  }
+
+  memset(bitset_, 0, num_bytes_);
+}
+
+BloomFilter::BloomFilter(const uint32_t* bitset, uint32_t num_bytes)
+    : pool_(::arrow::default_memory_pool()),
+	  bitset_(NULL),
+	  is_internal_bitset_(false),
+	  num_bytes_(num_bytes),
+      hash_strategy_(HashStrategy::MURMUR3_X64_128),
+      algorithm_(Algorithm::BLOCK) {
+  switch (hash_strategy_) {
+    case HashStrategy::MURMUR3_X64_128:
+      this->hasher_.reset(new MurmurHash3());
+      break;
+    default:
+      throw parquet::ParquetException("Unsupported hash strategy");
+  }
+
+  switch (algorithm_) {
+  case Algorithm::BLOCK:
+	  this->bloom_algorithm_.reset(new BlockBasedAlgorithm());
+	  break;
+  default:
+	  throw parquet::ParquetException("Unsupported Bloom filter algorithm");
+  }
+
+  if (!bitset) {
+	  throw parquet::ParquetException("bitset pointer is NULL");
+  }
+
+  if (num_bytes < MINIMUM_BLOOM_FILTER_BYTES ||
+		  num_bytes > MAXIMUM_BLOOM_FILTER_BYTES ||
+		  (num_bytes & (num_bytes -1)) != 0) {
+    throw parquet::ParquetException("Given length of bitset is illegal");
+  }
+
+  if ((reinterpret_cast<std::uintptr_t>(bitset) & (X64_CACHE_ALIGNMENT - 1)) == 0) {
```

Comment:
Yes, it looks a bit weird. How about to make bitset_ to std::shared_ptr and accept std::shared_ptr here?

Reply:
I don't think that's the right choice here. shared_ptr very often is covering up for other design flaws.

The first step here is to understand what interface the callers need. I asked about this in the last review, but did not receive a reply.

Also, very importantly: this block still leaks memory by providing the caller no information about whether it can free the block it passed it.

Reply:
I haven't implemented both read/write side for parquet-cpp, while in parquet-mr read side there is  one place to call the BloomFilter(uint32_t* bitset, uint32_t len) to construct a Bloom filter from the ColumnChunkMetaData.